### PR TITLE
add fp8 to distributed checkpointing tests

### PR DESCRIPTION
**Pull request by pstjohn**
fsdp2 w/ fp8 checkpointing is currently failing on my local machine

```
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_ddp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                      [  4%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_ddp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                       [  9%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_ddp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                       [ 13%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_ddp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                        [ 18%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_mfsdp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                    [ 22%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_mfsdp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                     [ 27%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_mfsdp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                     [ 31%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_mfsdp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                      [ 36%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_fsdp2[use_fp8] FAILED                                                                                                                                                                                                                                                                                                                                    [ 40%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_fsdp2[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                     [ 45%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_fsdp2[use_fp8] FAILED                                                                                                                                                                                                                                                                                                                                     [ 50%]
tests/test_distributed_checkpointing.py::test_checkpoint_save_and_load_two_processes_fsdp2[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                      [ 54%]
tests/test_distributed_checkpointing.py::test_final_model_save_ddp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                             [ 59%]
tests/test_distributed_checkpointing.py::test_final_model_save_ddp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                              [ 63%]
tests/test_distributed_checkpointing.py::test_final_model_save_mfsdp[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                           [ 68%]
tests/test_distributed_checkpointing.py::test_final_model_save_mfsdp[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                            [ 72%]
tests/test_distributed_checkpointing.py::test_final_model_save_fsdp2[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                           [ 77%]
tests/test_distributed_checkpointing.py::test_final_model_save_fsdp2[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                            [ 81%]
tests/test_distributed_checkpointing.py::test_scheduler_resume_single_gpu[use_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                      [ 86%]
tests/test_distributed_checkpointing.py::test_scheduler_resume_single_gpu[no_fp8] PASSED                                                                                                                                                                                                                                                                                                                                                       [ 90%]
tests/test_distributed_checkpointing.py::test_scheduler_resume_two_gpu[use_fp8] FAILED                                                                                                                                                                                                                                                                                                                                                         [ 95%]
tests/test_distributed_checkpointing.py::test_scheduler_resume_two_gpu[no_fp8] PASSED
```

### CODE REVIEW

File: bionemo-recipes/models/esm2/tests/test_distributed_fp8.py
Comment (pstjohn) [inline bionemo-recipes/models/esm2/tests/test_distributed_fp8.py:1]: i'm not sure we need this test per se, but this was sortof for my own sanity to ensure these were being synced

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py [hunk 1]
```diff
@@ -268,6 +273,66 @@ def save_final_model_mfsdp(
 # ============================================================================
 
 
+class FP8ExtraStateLoadPlanner(DefaultLoadPlanner):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/checkpoint.py:276]: Ugh. We could consider upstreaming this to TE?

File: bionemo-recipes/models/esm2/tests/test_distributed_fp8.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    model = NVEsmForMaskedLM(config)
+    model = model.to(torch.cuda.current_device())
+    model = torch.nn.parallel.DistributedDataParallel(model)
+
+    model.train()
+
+    generator = torch.Generator()
+    generator.manual_seed(torch.distributed.get_rank())
+
+    fp8_recipe = DelayedScaling(fp8_format=Format.HYBRID, amax_history_len=16, amax_compute_algo="max")
```
Comment (jstjohn) [inline bionemo-recipes/models/esm2/tests/test_distributed_fp8.py:63]: While I'm onboard with making sure everything works, given that we know that there can be an accuracy issue with delayed scaling, should we make the default recipes use what we think users should use instead? Maybe for example a check if blackwell or newer is available, then MXFP8, otherwise Block?

Comment (jstjohn) [reply bionemo-recipes/models/esm2/tests/test_distributed_fp8.py:63]: Or at least a comment about how DelayedScaling has accuracy issues on hopper, and direct users to the deepseekv3 paper for more details?

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py (hunk 1 above)
Comment (jstjohn) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py:276]: Curious what kind of extra state is needed for the other fp8 recipes... They may simplify this?

File: bionemo-recipes/models/esm2/tests/test_distributed_fp8.py (hunk 2 above)
Comment (jstjohn) [reply bionemo-recipes/models/esm2/tests/test_distributed_fp8.py:63]: Oh right this is a test file. What's the default in the main recipe? Apply this suggestion to that.

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py (hunk 1 above)
Comment (jstjohn) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py:276]: We could just recommend users not use this recipe if it adds complexity, and point to the accuracy issue for background on why we aren't supporting it as well as the complexity, however I'm not sure whether the other recipes need to save extra state or not...

Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py:276]: It seems i still need this even for `Float8CurrentScaling`. I wonder if that `_extra_state` is actually doing anything though

File: bionemo-recipes/models/esm2/tests/test_distributed_fp8.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/tests/test_distributed_fp8.py:63]: Yeah i change it now in defaults.yaml to Float8CurrentScaling -- i could definitely make that a `null` and do some runtime inference of an optimal recipe; but i think the issue here (fp8 data requiring custom dcp handling) is still unsolved.

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py [hunk 3]
```diff
@@ -333,7 +398,8 @@ def load_checkpoint_fsdp2(
     app_state = AppState(model=model, optimizer=optimizer, scheduler=scheduler)
 
     state_dict = {"app": app_state}
-    dcp.load(state_dict, checkpoint_id=checkpoint_path)
+    # Use custom planner that handles FP8 _extra_state shape mismatches
+    dcp.load(state_dict, checkpoint_id=checkpoint_path, planner=FP8ExtraStateLoadPlanner())
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/esm2_native_te/checkpoint.py:402]: IS it ok to have this planner if we aren't using FP8

File: bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py [hunk 4]
```diff
@@ -187,7 +204,10 @@ def test_checkpoint_save_and_load_two_processes_ddp(recipe_path, tmp_path):
         assert expected in final_checkpoint_files, f"Missing checkpoint: {expected}"
 
 
-def test_checkpoint_save_and_load_single_process_mfsdp(recipe_path, tmp_path):
+@pytest.mark.parametrize(
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py:207]: Can we just make another test instead of having this parametrize thing? We can just make one that covers the FP8 case instead? This way if FP8 is failing the other can still pass etc.

File: bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py [hunk 5]
```diff
@@ -257,7 +279,10 @@ def test_checkpoint_save_and_load_single_process_mfsdp(recipe_path, tmp_path):
 
 
 @requires_multi_gpu
-def test_checkpoint_save_and_load_two_processes_mfsdp(recipe_path, tmp_path):
+@pytest.mark.parametrize(
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py:282]: more of what I said above about making a new test for FP8 coverage.

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py (hunk 3 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py:402]: yup. we have a test for that with no FP8, it should just behave like the default planner

File: bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py (hunk 4 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/tests/test_distributed_checkpointing.py:207]: that happens here; it essentially looks like
```
...
bionemo-test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_fsdp2[use_fp8]
bionemo-test_distributed_checkpointing.py::test_checkpoint_save_and_load_single_process_fsdp2[no_fp8]
...
```
where each parameterized variant is a separate test

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py (hunk 1 above)
Comment (cspades) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py:276]: @pstjohn This is cool but also couldn't get it to work on the Megatron-FSDP side, at least not in a way where I think it's maintainable for production code, i.e. institutionalized in a hook or something.

For now, I think Megatron-FSDP will require `strict=False` state dict loading when using TE models as well, as we do not support FP8 extra state in our checkpoints without the extra state sizing error (either 0 vs. non-zero or N vs. M in some cases) and the model needs to ignore the missing keys.
